lightning/src/ln/channel.rs

```diff
@@ -1691,7 +1704,8 @@ impl<Signer: Sign> Channel<Signer> {
 
 	pub fn funding_locked(&mut self, msg: &msgs::FundingLocked) -> Result<(), ChannelError> {
 		if self.channel_state & (ChannelState::PeerDisconnected as u32) == ChannelState::PeerDisconnected as u32 {
-			return Err(ChannelError::Close("Peer sent funding_locked when we needed a channel_reestablish".to_owned()));
+			self.workaround_lnd_bug_4006 = Some(msg.clone());
+			return Err(ChannelError::Ignore("Peer sent funding_locked when we needed a channel_reestablish. The peer is likely lnd, see their bug #4006.".to_owned()));
```

Comment:
Won't this mean any implementation of `ChannelMessageHandler` needs to implement this workaround. Not sure if there is a better way, though. Maybe cache it at the peer handling layer using a wrapper around `ChannelMessageHandler`? Might not be feasible if you need the channel state. :/

Reply:
Yes, if you implement a whole new channel implementation then you would, but I'm not sure how likely that is. You wouldn't need to if you were implementing `ChannelMessageHandler` as a simple wrapper around our `ChannelManager` (which I do expect people to do).